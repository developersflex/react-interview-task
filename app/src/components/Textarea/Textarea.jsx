import React from 'react';
import './Textarea.scss';
export const Textarea = ({ className = '', ...props }) => {
  return <textarea className={`kit-textarea ${className}`} {...props} />;
};
