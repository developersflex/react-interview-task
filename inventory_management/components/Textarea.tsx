import { useField } from "formik";
import React from "react";

type TextareaProps = {
  placeholder: string;
  label: string;
  name: string;
  className?: string;
  rows?: number;
  cols?: number;
};

const Textarea: React.FC<TextareaProps> = ({
  placeholder,
  label,
  name,
  rows = 5,
  cols = 5,
  className,
}) => {
  const [field, meta] = useField(name);
  const hasError = meta.touched && meta.error;

  return (
    <div className={`w-full mb-5 ${className}`}>
      <div className="relative flex flex-col gap-1">
        <label className="ml-4 text-base font-semibold">{label}</label>
        <textarea
          id={label}
          rows={rows}
          cols={cols}
          placeholder={placeholder}
          className="rounded-md bg-brand-background-primary px-2 py-1"
          {...field}
        />
        {hasError && (
          <p className="text-brand-red-1000 text-xs ml-2 absolute -bottom-5">
            {meta.error}
          </p>
        )}
      </div>
    </div>
  );
};

export default Textarea;
