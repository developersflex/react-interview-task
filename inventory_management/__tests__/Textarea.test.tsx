import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { Formik, Form } from "formik";
import Textarea from "@/components/Textarea";
import * as Yup from "yup";

describe("Textarea Component", () => {
  it("renders without errors", () => {
    const { container } = render(
      <Formik initialValues={{ testName: "" }} onSubmit={() => {}}>
        <Textarea
          placeholder="Test Placeholder"
          label="Test Label"
          name="testName"
        />
      </Formik>
    );
    expect(container).toMatchSnapshot();
  });

  it("displays the label and placeholder correctly", () => {
    const { getByLabelText, getByPlaceholderText } = render(
      <Formik initialValues={{ testName: "" }} onSubmit={() => {}}>
        <Textarea
          placeholder="Test Placeholder"
          label="Test Label"
          name="testName"
        />
      </Formik>
    );

    expect(getByLabelText("Test Label")).toBeInTheDocument();
    expect(getByPlaceholderText("Test Placeholder")).toBeInTheDocument();
  });

  it("updates field value on user input", async () => {
    const { getByPlaceholderText } = render(
      <Formik initialValues={{ testName: "" }} onSubmit={() => {}}>
        <Textarea
          placeholder="Test Placeholder"
          label="Test Label"
          name="testName"
        />
      </Formik>
    );

    const textarea = getByPlaceholderText("Test Placeholder");
    fireEvent.change(textarea, { target: { value: "Updated Value" } });

    await waitFor(async () => {
      expect(textarea).toHaveValue("Updated Value");
    });
  });

  it("displays error message when there are validation errors", async () => {
    const validationSchema = Yup.object().shape({
      testName: Yup.string().required("This field is required."),
    });

    const { getByText } = render(
      <Formik
        initialValues={{ testName: "" }}
        onSubmit={() => {}}
        validationSchema={validationSchema}
      >
        <Form>
          <Textarea
            placeholder="Test Placeholder"
            label="Test Label"
            name="testName"
          />
          <button type="submit">Submit</button>
        </Form>
      </Formik>
    );

    fireEvent.click(getByText("Submit"));

    await waitFor(async () => {
      const expectedErrorMessage = "This field is required.";
      expect(getByText(expectedErrorMessage)).toBeInTheDocument();
    });
  });
});
