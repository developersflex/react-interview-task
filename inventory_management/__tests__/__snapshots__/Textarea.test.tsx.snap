// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Component renders without errors 1`] = `
<div>
  <div
    class="w-full mb-5 undefined"
  >
    <div
      class="relative flex flex-col gap-1"
    >
      <label
        class="ml-4 text-base font-semibold"
        for="testName"
      >
        Test Label
      </label>
      <textarea
        class="rounded-md bg-brand-background-primary px-2 py-1"
        cols="5"
        id="testName"
        name="testName"
        placeholder="Test Placeholder"
        rows="5"
      />
    </div>
  </div>
</div>
`;
